### ISSUE TITLE
Adding a transformer to an array of objects

### ISSUE THREAD

**Issue by pedrem** [jolicode/automapper#111]
Hello,

I'm trying to transform an array of objects (DTOs) into an array(entities), Is that something that the library is capable of handling?

Here's a snippet of the code:

```
/** @var ColourApi[] */
#[MapTo(transformer: ColourTransformer::class)]
public array $colours = [];
```

When looking at the automapper tab in the profiler although it attaches AutoMapper\Transformer\PropertyTransformer\PropertyTransformer	there's no code inside.

I've the same setup for simple properties, so I know the problem is probably not the transformer. 

Let me know if you need more context.

Thank you!

**Comment by joelwurtz** [comment]
It should not be a problem normally, we may need more or a simple reproducer so we can see what's the problem here

**Comment by pedrem** [comment]
I'll try to provide a little more context:

So here's a snippet of the entity which I'm trying to map too:

```    
/** @var Colour[] */
#[ORM\ManyToMany(targetEntity: Colour::class)]
#[ORM\JoinTable(name: 'campaignProductColours')]
#[ORM\JoinColumn(name: 'campaignProductId', referencedColumnName: 'campaignProductId')]
private Collection $colours;
```
Without adding a custom transformer it uses `ArrayTransformer`, but once I add the transformer(1st snippet) no code is actually attached, see example image below:

No custom transformer on colours property
![Screenshot 2024-04-17 at 09 47 29](https://github.com/jolicode/automapper/assets/22331228/6cf5aab1-7f96-46e7-b11c-ca81fcbc8bc0)

Custom transformer on colours property
![Screenshot 2024-04-17 at 09 45 32](https://github.com/jolicode/automapper/assets/22331228/98ace8cf-c0b7-4e15-a74b-08c1b7ae6612)

I've tried both these approaches but same result:
- https://automapper.jolicode.com/9.0.0-beta.2/mapping/transformer/#creating-a-custom-transformer
- https://automapper.jolicode.com/9.0.0-beta.2/mapping/transformer/#automatically-apply-custom-transformers


Bottom line I'm trying to hydrate the entities using the transformer, so please let me know if there's a way without using transformers!

**Comment by joelwurtz** [comment]
Does the `ColourTransformer` is registered as a service (under the id `ColourTransformer::class` ?

**Comment by joelwurtz** [comment]
> Bottom line I'm trying to hydrate the entities using the transformer, so please let me know if there's a way without using transformers!

Yes you could use a provider see https://automapper.jolicode.com/dev/mapping/provider/

**Comment by pedrem** [comment]
> Does the `ColourTransformer` is registered as a service (under the id `ColourTransformer::class` ?

Yeah, you can see on the 2nd screenshot it correctly attaches to `colourFeatured`.

Not really sure this is the "problem" but I've been trying to understand why the code is not getting attached and found that it might be related to [this](https://github.com/jolicode/automapper/blob/aa27772b07053c553bf87593378e87731b1346e9/src/Extractor/WriteMutator.php#L108).

**Comment by joelwurtz** [comment]
Not exactly but something close to this: this may be because there is a transformer on a collection but the hydrate part can only be done with add / remove method.

I will try to look at that, can add a 'setColours' on the target to see if it works ? (or make the property public)

**Comment by pedrem** [comment]
> I will try to look at that, can you add a 'setColours' on the target to see if it works ? (or make the property public)

Didn't seem to make a difference

**Comment by joelwurtz** [comment]
Yeah, for the record, provider was badly injected in the bundle (should be fixed in next version).

For adder and remove i made https://github.com/jolicode/automapper/pull/116 which should resolve your first issue.

For the record, it will transform the array entirely, transformer will not be applied to the item (so it need to handle an array as an input and return an array). Then it will correctly used the adder method

I think it's better than the item, as it will allow, in your case, to fetch a collection instead of doing a sql query for each item, also this is consistent with other way to write the values.
